File: exercism/java/anagram/src/main/java/Anagram.java
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Anagram{
    private Map<Integer, Long> countingMap;
    private String word;
    public Anagram(String word){
        this.word = word.toLowerCase();
        this.countingMap = toCountingMap(word);
    }

    private Map<Integer, Long> toCountingMap(String word){
        return word.toLowerCase().chars()
                .mapToObj(Integer::valueOf)
                .collect(
                        Collectors.groupingBy(
                                Function.identity(),
                                Collectors.counting()
                        )
                );
    }
        
    public List<String> match(List<String> list){
        return list.stream()
                .filter(
                        i -> !this.word.equalsIgnoreCase(i)
                                && toCountingMap(i).equals(countingMap)
                )
                .collect(Collectors.toList());
    }

    public static void main(String[] args) {
        new Anagram("test");
    }
}

File: exercism/kotlin/anagram/src/main/kotlin/Anagram.kt
class Anagram(private val source: String) {

    private val sourceMap = source.countMap()
    private val sourceLower = source.toLowerCase()

    fun <T>Collection<T>.countBy(): Map<T,Int> = this
            .groupBy { it }
            .map { it.key to it.value.size }.toMap()

    fun String.countMap(): Map<Char,Int> = this.toLowerCase().toList().countBy()

    fun match(anagrams: Collection<String>): Set<String> {
            return anagrams.filter { it.toLowerCase() != sourceLower && it.countMap() == sourceMap }.toSet()
    }
}
